### ISSUE TITLE
add operation point dependent efficiciency to processes

### ISSUE THREAD

**Issue by ojdo** [tum-ens/urbs#1]
Instead of a flat efficiency (attribute `eff`), conversion processes could have operation dependent efficiency, i.e. depending on the fraction `tau_pro`/`cap_pro`. 

Implementation ideas: There is a way to formulate this feature using completely linearly (LP compatible) using a continuous variable that reflects the "online capacity", e.g. `cap_online`, that may move freely between `0` and `cap_pro`. Changes of `cap_online` are prohibited using an additional ramp parameter, that reflects startup costs.

**Comment by ojdo** [comment]
Now that I found my notes, here is a possible formulation in pseudo-code (and possibly some rough edges). Needed are the new time-dependent variables `cap_online` and `cap_start` for each process and the new attribute `partial` [unitless fraction, from 0 to 1] that enforces partial load capability of the online capacity.

    cap_online[t] * partial <= tau_pro[t] <= cap_online[t]
    cap_start[t] >= cap_online[t] - cap_online[t-1]
    cap_start[t] >= 0

What this does: the first inequality limits the process throughput to be in the interval between `cap_online * partial` and `cap_online`.

Cost function then needs to be extended using the startup cost term and the online capacity:

    z += sum(c_start * cap_start[t] + c_online * cap_online[t] for t in time)

Finally, the output/input fractions can depend on both `tau_pro` and `cap_online`:

    e_pro_out = tau_pro * r_out_var + cap_online * r_out_fix

I'm still unsure whether this formulation or another (using something like `eff-min`, `eff-max`) might be better. This still needs some pen&paper-work before being "ready to code".

So all in all this sketch includes roughly the features:

  * Startup costs
  * Ramping constraints (by limiting `cap_start`)
  * Partial load capability (attribute `partial`)
  * Operation dependent efficiencies (using `r_out_var` and `r_out_fix`)

**Comment by sonercandas** [comment]
Staying on the 'binary variable' side of handling the issue, another way of approach might be using the following sets of inequalities to construct an "either zero or a bounded throughput", via introduction of `onstatus` which is the binary variable assuming the value 0 or 1 according to the plant's running-or-not situation, the auxiliary & 'non-negative defined' variable `aux` (placeholder name) and parameter `partial` which is input by the user to define the minimum fraction of the plant's part load operation:

```
aux[t] <= cap_pro
aux[t] <= N * m.onstatus[t]
aux[t] >= cap_pro - N * (1 - onstatus[t])
% N is an arbitrary number much larger than cap_pro~
```
with the addition of the following constraint:

`
(partial * aux[t], tau_pro[t], aux[t])
`

Former set of inequalities constrains the variable `aux` to be equal to zero whenever `onstatus `is 0, and equal to `cap_pro` whenever `onstatus` is 1. 

The latter inequality basically confines `tau_pro` to 0 whenever `aux` (hence, `onstatus`) is 0 and sets the desired range of values (from partial * capacity to capacity) whenever `aux` (hence, `onstatus`) is non-zero. **Note:** The rule `res_process_throughput_by_capacity_rule` gets obsolete with the addition of this constraint.

This method seemingly takes care of the minimum load fraction feature. To implement the startup costs `c_start` for each timestep `t` , we could check `onstatus[t-1] == 0 and onstatus[t] == 1` (the timestep at which plant starts its operation) via introduction of a new binary variable `aux2` and the following set of inequalities (this structure is easier since both variables are binary and somewhat simpler to play with):

```
aux2[t] <= onstatus[t]
aux2[t] >= onstatus[t] - 2 * onstatus[t-1]
aux2[t] <= (3 * onstatus[t] - onstatus[t-1] + 1) / 4
```

such that `aux2[t]` only only gets value of 1 when `onstatus[t-1] == 0 and onstatus[t] == 1` and 0 for every other case. We could then simply:

`z += sum(aux2[t] * c_start for t in time)`

**Remark:** Addition of variable efficiencies might take even more computational effort since its equations seemingly contain more non-linearity than the start-up and partial load ones. The model got quite cumbersome already with only the partial load feature in my system.

**Comment by sonercandas** [comment]
My commit for above mentioned changes are as follows; with some comments and a simple 24-timestep- input file. https://github.com/sonercandas/urbs/commit/0a1083b1c675ad5fa75a29182e18bfb0edeb771f

**Remarks:** 
While the partial load feature works as intended; it seemingly intrudes the "still-active" maximum gradient rule at start-up timesteps e.g. the throughput can jump from `0` to `partial * cap_pro` which doesn't make sense since the gradient is constraint. 
Also, the start-up cost component is, in the current state, independent of the process capacity; which does not really reflect the reality. Addition of capacity dependency introduces a non-linearity which might be solved with another auxiliary variable/s.

**Comment by ojdo** [comment]
Here is the full "solution", LP partial load efficiencies and startup costs. 

### Status quo

The default [process constraints](http://urbs.readthedocs.io/en/latest/mathdoc/equations.html#process-constraints) with their parameters [input and output ratios](http://urbs.readthedocs.io/en/latest/mathdoc/parameters.html#process-technical-parameters) look like this:

    e_pro_in = tau_pro * r_in
    e_pro_out = tau_pro * r_out

Note that the conversion from in to out has the efficiency `e_pro_out/e_pro_in = r_out/r_in`. In other words: an energy conversion process of 40% efficiency can either be expressed by `r_in=1, r_out=0.4` or `r_in=2.5, r_out=1` or any other pair of numbers whose fraction is 0.4. This gives a degree of freedom to what the meaning of variable `tau_pro`. For the following, I will assume `r_out = 1` and thus `tau_pro = e_pro_out` (e.g. electricity production of a plant, if you will). We will focus only on the first constraint, and modify it to lead to a non-constant efficiency over the operational range of the process. This range is first only limited by the total process capacity `kappa`:

    0 <= tau_pro <= kappa

### New formulation

Two additional, time-dependent process state variables are introduced. The first `cap_online` (MW) is. Additionally, there is a startup indicator variable `start_cap` (MW), always non-zero whenever `cap_online` is increasing.

The input ratio parameter `r_in` is split into two values `r` (lower input ratio) and `R` (upper input ratio). A minimum partial load parameter `a` is also introduced.

    min_fraction * cap_online <= tau_pro <= cap_online
    cap_online <= cap_pro
    start_cap >= (cap_online[t] - cap_online[t-1]) / dt    [dt involvement must be clarified]
    e_pro_in = cap_online * min_fraction * (r - R) / (1 - min_fraction)
               + tau_pro * (R - min_fraction * r) / (1 - min_fraction)
    costs += sum[t in Tm] k_start * start_cap[t] * weight

The new definition for `e_pro_in` creates an effectively non-linear input-output efficiency, as plotted [here](http://nbviewer.jupyter.org/gist/ojdo/a8d26dc6abc9d22faf77d7cd2623dddc/startup-partial-ratios.ipynb).

**Comment by ojdo** [comment]
Not yet ready for a pull request, but branch [`startupandpartial`](https://github.com/ojdo/urbs/tree/startupandpartial) (based on @sonercandas identically named PR/branch) now implements this new formulation. Once I have refreshed mathdoc, I will proceed to PR'ing.

**Comment by ojdo** [comment]
Closed by #70.
